libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,191 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import timber.log.Timber;
+
+class EnhancedLocationEngineDecorator implements LocationEngine {
+  private static final long UPDATE_INTERVAL_IN_MILLISECONDS = 1000;
+  private static final long FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS = 500;
+  private final LocationEngine locationEngine;
+  private final MapboxNavigator mapboxNavigator;
+  private final LocationEngineCallback<LocationEngineResult> callback = new CurrentLocationEngineCallback(this);
+  private ScheduledFuture future;
+  private final ScheduledExecutorService executorService;
+  private Location rawLocation;
+  private Location enhancedLocation;
+
+  EnhancedLocationEngineDecorator(@NonNull LocationEngine locationEngine,
+                                  @NonNull MapboxNavigator mapboxNavigator,
+                                  @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, callback, looper);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     PendingIntent pendingIntent) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, pendingIntent);
+  }
+
+  @Override
+  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @Override
+  public void removeLocationUpdates(PendingIntent pendingIntent) {
+    locationEngine.removeLocationUpdates(pendingIntent);
+  }
+
+  void onActiveGuidance() {
+    removeLocationUpdates(callback);
+    if (future != null) {
+      future.cancel(false);
+      future = null;
+    }
+  }
+
+  void onFreeDrive() {
+    requestLocationUpdates(obtainLocationEngineRequest(), callback, null);
+    final Handler handler = new Handler(Looper.getMainLooper());
+    future = executorService.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        if (rawLocation != null) {
+          enhancedLocation = getLocation(new Date(), 0, rawLocation);
+          handler.post(new Runnable() {
+            @Override
+            public void run() {
+              callback.onSuccess(LocationEngineResult.create(enhancedLocation));
```

Comment:
The decorator currently doesn't provide any data outside, it just feeds itself in a loop while the user is provided with raw location updates only.

`callback` invoked here should be **the original callback that the user-provided** otherwise we're just looping the location updates.

Additionally, the original callback should be unregistered, when free-drive mode starts, otherwise, the original listener will receive two updates, one with raw location and one with enhanced one.

Reply:
good catch!